File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/Json.java
package spectable.common;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import production.*;
import records.*;
import calculator.*;

/** Minimal dependency-free JSON reader/writer used by the generated Typed classes. */
public final class Json {

    private Json() {}

    // -----------------------------------------------------------------
    // Writing
    // -----------------------------------------------------------------

    /** Serialize a value graph (null, String, Boolean, Number, Map, Iterable) to JSON text. */
    public static String write(Object value) {
        StringBuilder sb = new StringBuilder();
        writeValue(sb, value);
        return sb.toString();
    }

    private static void writeValue(StringBuilder sb, Object v) {
        if (v == null)               { sb.append("null"); return; }
        if (v instanceof String)     { writeString(sb, (String) v); return; }
        if (v instanceof Character)  { writeString(sb, String.valueOf(v)); return; }
        if (v instanceof Boolean)    { sb.append(((Boolean) v).booleanValue() ? "true" : "false"); return; }
        if (v instanceof BigDecimal) { sb.append(((BigDecimal) v).toPlainString()); return; }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            if (Double.isNaN(d) || Double.isInfinite(d))
                throw new IllegalArgumentException("Cannot write non-finite number to JSON: " + d);
            sb.append(Double.toString(d));
            return;
        }
        if (v instanceof Number) { sb.append(v.toString()); return; }
        if (v instanceof Map) {
            sb.append('{');
            boolean first = true;
            for (Map.Entry<?, ?> e : ((Map<?, ?>) v).entrySet()) {
                if (!first) sb.append(',');
                first = false;
                writeString(sb, String.valueOf(e.getKey()));
                sb.append(':');
                writeValue(sb, e.getValue());
            }
            sb.append('}');
            return;
        }
        if (v instanceof Iterable) {
            sb.append('[');
            boolean first = true;
            for (Object o : (Iterable<?>) v) {
                if (!first) sb.append(',');
                first = false;
                writeValue(sb, o);
            }
            sb.append(']');
            return;
        }
        writeString(sb, v.toString());
    }

    private static void writeString(StringBuilder sb, String s) {
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\b': sb.append("\\b");  break;
                case '\f': sb.append("\\f");  break;
                case '\n': sb.append("\\n");  break;
                case '\r': sb.append("\\r");  break;
                case '\t': sb.append("\\t");  break;
                default:
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else          sb.append(c);
            }
        }
        sb.append('"');
    }

    // -----------------------------------------------------------------
    // Reading
    // -----------------------------------------------------------------

    /** Parse JSON text into Map / List / String / BigDecimal / Boolean / null. */
    public static Object parse(String text) {
        if (text == null) throw new IllegalArgumentException("JSON text is null");
        Parser p = new Parser(text);
        Object v = p.readValue();
        p.skipWhitespace();
        if (!p.atEnd()) throw new IllegalArgumentException("Trailing content at offset " + p.pos());
        return v;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseObject(String text) {
        Object v = parse(text);
        if (!(v instanceof Map)) throw new IllegalArgumentException("Expected a JSON object");
        return (Map<String, Object>) v;
    }

    @SuppressWarnings("unchecked")
    public static List<Object> parseArray(String text) {
        Object v = parse(text);
        if (!(v instanceof List)) throw new IllegalArgumentException("Expected a JSON array");
        return (List<Object>) v;
    }

    private static final class Parser {
        private final String src;
        private int i = 0;

        Parser(String src) { this.src = src; }

        int pos()        { return i; }
        boolean atEnd()  { return i >= src.length(); }

        void skipWhitespace() {
            while (i < src.length()) {
                char c = src.charAt(i);
                if (c == ' ' || c == '\t' || c == '\n' || c == '\r') i++;
                else break;
            }
        }

        Object readValue() {
            skipWhitespace();
            if (atEnd()) throw err("Unexpected end of JSON input");
            char c = src.charAt(i);
            switch (c) {
                case '{': return readObject();
                case '[': return readArray();
                case '"': return readString();
                case 't': expect("true");  return Boolean.TRUE;
                case 'f': expect("false"); return Boolean.FALSE;
                case 'n': expect("null");  return null;
                default:  return readNumber();
            }
        }

        Map<String, Object> readObject() {
            Map<String, Object> m = new LinkedHashMap<String, Object>();
            i++; // consume '{'
            skipWhitespace();
            if (!atEnd() && src.charAt(i) == '}') { i++; return m; }
            while (true) {
                skipWhitespace();
                if (atEnd() || src.charAt(i) != '"') throw err("Expected a string key");
                String key = readString();
                skipWhitespace();
                if (atEnd() || src.charAt(i) != ':') throw err("Expected ':' after key '" + key + "'");
                i++;
                m.put(key, readValue());
                skipWhitespace();
                if (atEnd()) throw err("Unterminated object");
                char d = src.charAt(i);
                if (d == ',') { i++; continue; }
                if (d == '}') { i++; return m; }
                throw err("Expected ',' or '}'");
            }
        }

        List<Object> readArray() {
            List<Object> list = new ArrayList<Object>();
            i++; // consume '['
            skipWhitespace();
            if (!atEnd() && src.charAt(i) == ']') { i++; return list; }
            while (true) {
                list.add(readValue());
                skipWhitespace();
                if (atEnd()) throw err("Unterminated array");
                char d = src.charAt(i);
                if (d == ',') { i++; continue; }
                if (d == ']') { i++; return list; }
                throw err("Expected ',' or ']'");
            }
        }

        String readString() {
            i++; // consume opening quote
            StringBuilder sb = new StringBuilder();
            while (true) {
                if (atEnd()) throw err("Unterminated string");
                char c = src.charAt(i++);
                if (c == '"')  return sb.toString();
                if (c != '\\') { sb.append(c); continue; }
                if (atEnd()) throw err("Unterminated escape");
                char e = src.charAt(i++);
                switch (e) {
                    case '"':  sb.append('"');  break;
                    case '\\': sb.append('\\'); break;
                    case '/':  sb.append('/');  break;
                    case 'b':  sb.append('\b'); break;
                    case 'f':  sb.append('\f'); break;
                    case 'n':  sb.append('\n'); break;
                    case 'r':  sb.append('\r'); break;
                    case 't':  sb.append('\t'); break;
                    case 'u':
                        if (i + 4 > src.length()) throw err("Truncated \\u escape");
                        sb.append((char) Integer.parseInt(src.substring(i, i + 4), 16));
                        i += 4;
                        break;
                    default: throw err("Invalid escape '\\" + e + "'");
                }
            }
        }

        BigDecimal readNumber() {
            int start = i;
            if (!atEnd() && src.charAt(i) == '-') i++;
            while (!atEnd()) {
                char c = src.charAt(i);
                if ((c >= '0' && c <= '9') || c == '.' || c == 'e' || c == 'E'
                 || c == '+' || c == '-') i++;
                else break;
            }
            if (start == i) throw err("Expected a value");
            String text = src.substring(start, i);
            try {
                return new BigDecimal(text);
            } catch (NumberFormatException ex) {
                throw err("Invalid number '" + text + "'");
            }
        }

        void expect(String word) {
            if (!src.startsWith(word, i)) throw err("Expected '" + word + "'");
            i += word.length();
        }

        IllegalArgumentException err(String msg) {
            return new IllegalArgumentException(msg + " at offset " + i);
        }
    }

    // -----------------------------------------------------------------
    // Field accessors — a missing key or a wrong type raises
    // IllegalArgumentException.  An explicit JSON null is passed through
    // as null (for reference types) rather than treated as an error.
    // -----------------------------------------------------------------

    public static Object require(Map<String, Object> obj, String key) {
        if (obj == null)              throw new IllegalArgumentException("Missing JSON object for field '" + key + "'");
        if (!obj.containsKey(key))    throw new IllegalArgumentException("Missing JSON field '" + key + "'");
        return obj.get(key);
    }

    public static String              getString (Map<String, Object> o, String k) { return asString (require(o, k), k); }
    public static int                 getInt    (Map<String, Object> o, String k) { return asInt    (require(o, k), k); }
    public static long                getLong   (Map<String, Object> o, String k) { return asLong   (require(o, k), k); }
    public static double              getDouble (Map<String, Object> o, String k) { return asDouble (require(o, k), k); }
    public static BigDecimal          getDecimal(Map<String, Object> o, String k) { return asDecimal(require(o, k), k); }
    public static boolean             getBoolean(Map<String, Object> o, String k) { return asBoolean(require(o, k), k); }
    public static char                getChar   (Map<String, Object> o, String k) { return asChar   (require(o, k), k); }
    public static Map<String, Object> getObject (Map<String, Object> o, String k) { return asObject (require(o, k), k); }
    public static List<Object>        getArray  (Map<String, Object> o, String k) { return asArray  (require(o, k), k); }

    public static String asString(Object v, String ctx) {
        if (v == null)               return null;
        if (v instanceof String)     return (String) v;
        if (v instanceof BigDecimal) return ((BigDecimal) v).toPlainString();
        if (v instanceof Boolean)    return v.toString();
        throw typeError(ctx, "a string", v);
    }

    public static BigDecimal asDecimal(Object v, String ctx) {
        if (v instanceof BigDecimal) return (BigDecimal) v;
        if (v instanceof Number)     return new BigDecimal(v.toString());
        if (v instanceof String) {
            try { return new BigDecimal(((String) v).trim()); }
            catch (NumberFormatException ex) { throw typeError(ctx, "a number", v); }
        }
        throw typeError(ctx, "a number", v);
    }

    public static int asInt(Object v, String ctx) {
        try { return asDecimal(v, ctx).stripTrailingZeros().intValueExact(); }
        catch (ArithmeticException ex) { throw typeError(ctx, "an integer", v); }
    }

    public static long asLong(Object v, String ctx) {
        try { return asDecimal(v, ctx).stripTrailingZeros().longValueExact(); }
        catch (ArithmeticException ex) { throw typeError(ctx, "a long", v); }
    }

    public static double asDouble(Object v, String ctx) {
        return asDecimal(v, ctx).doubleValue();
    }

    public static boolean asBoolean(Object v, String ctx) {
        if (v instanceof Boolean) return ((Boolean) v).booleanValue();
        if (v instanceof String) {
            String s = ((String) v).trim();
            if (s.equalsIgnoreCase("true")  || s.equalsIgnoreCase("t")
             || s.equalsIgnoreCase("yes")   || s.equalsIgnoreCase("y")  || s.equals("1")) return true;
            if (s.equalsIgnoreCase("false") || s.equalsIgnoreCase("f")
             || s.equalsIgnoreCase("no")    || s.equalsIgnoreCase("n")  || s.equals("0")) return false;
        }
        throw typeError(ctx, "a boolean", v);
    }

    public static char asChar(Object v, String ctx) {
        String s = asString(v, ctx);
        return (s == null || s.isEmpty()) ? '\0' : s.charAt(0);
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> asObject(Object v, String ctx) {
        if (v == null)        return null;
        if (v instanceof Map) return (Map<String, Object>) v;
        throw typeError(ctx, "an object", v);
    }

    @SuppressWarnings("unchecked")
    public static List<Object> asArray(Object v, String ctx) {
        if (v == null)         return null;
        if (v instanceof List) return (List<Object>) v;
        throw typeError(ctx, "an array", v);
    }

    private static IllegalArgumentException typeError(String ctx, String expected, Object actual) {
        return new IllegalArgumentException(
            "JSON field '" + ctx + "' is not " + expected + " (got " + describe(actual) + ")");
    }

    private static String describe(Object v) {
        if (v == null)               return "null";
        if (v instanceof String)     return "a string";
        if (v instanceof Boolean)    return "a boolean";
        if (v instanceof BigDecimal) return "a number";
        if (v instanceof Map)        return "an object";
        if (v instanceof List)       return "an array";
        return v.getClass().getSimpleName();
    }
}

File: Rust/ExampleTests/common/json.rs
//! Minimal dependency-free JSON reader/writer used by the generated Typed structs.
//!
//! No external crate is required. A missing field or a value of the wrong type
//! produces `Err(JsonError)`; an explicit JSON null is passed through.

#![allow(dead_code)]

use std::collections::BTreeMap;
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub struct JsonError {
    pub message: String,
}

impl JsonError {
    pub fn new(message: impl Into<String>) -> Self {
        Self { message: message.into() }
    }
}

impl fmt::Display for JsonError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for JsonError {}

pub type JsonResult<T> = Result<T, JsonError>;

/// A parsed JSON value. Numbers keep their source text so that decimal and
/// long integer fields survive a round trip exactly.
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Number(String),
    Str(String),
    Array(Vec<Value>),
    Object(Vec<(String, Value)>),
}

impl Value {
    pub fn number_from_i64(v: i64) -> Value { Value::Number(v.to_string()) }

    pub fn number_from_f64(v: f64) -> Value {
        if v.is_finite() {
            // {:?} gives the shortest representation that round-trips.
            Value::Number(format!("{:?}", v))
        } else {
            Value::Null
        }
    }

    pub fn get(&self, key: &str) -> Option<&Value> {
        match self {
            Value::Object(members) => members.iter().find(|(k, _)| k == key).map(|(_, v)| v),
            _ => None,
        }
    }

    pub fn describe(&self) -> &'static str {
        match self {
            Value::Null      => "null",
            Value::Bool(_)   => "a boolean",
            Value::Number(_) => "a number",
            Value::Str(_)    => "a string",
            Value::Array(_)  => "an array",
            Value::Object(_) => "an object",
        }
    }
}

/// Kept so callers that prefer a map view have one; the parser preserves order.
pub fn to_map(value: &Value) -> BTreeMap<String, Value> {
    match value {
        Value::Object(members) => members.iter().cloned().collect(),
        _ => BTreeMap::new(),
    }
}

// ---------------------------------------------------------------------
// Writing
// ---------------------------------------------------------------------

pub fn escape_into(out: &mut String, s: &str) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"'  => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0c}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
}

pub fn write_into(out: &mut String, value: &Value) {
    match value {
        Value::Null      => out.push_str("null"),
        Value::Bool(b)   => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => out.push_str(n),
        Value::Str(s)    => escape_into(out, s),
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 { out.push(','); }
                write_into(out, item);
            }
            out.push(']');
        }
        Value::Object(members) => {
            out.push('{');
            for (i, (k, v)) in members.iter().enumerate() {
                if i > 0 { out.push(','); }
                escape_into(out, k);
                out.push(':');
                write_into(out, v);
            }
            out.push('}');
        }
    }
}

pub fn write(value: &Value) -> String {
    let mut out = String::new();
    write_into(&mut out, value);
    out
}

// ---------------------------------------------------------------------
// Reading
// ---------------------------------------------------------------------

struct Parser<'a> {
    src: &'a [u8],
    text: &'a str,
    i: usize,
}

impl<'a> Parser<'a> {
    fn new(text: &'a str) -> Self {
        Self { src: text.as_bytes(), text, i: 0 }
    }

    fn err<T>(&self, msg: &str) -> JsonResult<T> {
        Err(JsonError::new(format!("{} at offset {}", msg, self.i)))
    }

    fn at_end(&self) -> bool { self.i >= self.src.len() }

    fn skip_ws(&mut self) {
        while self.i < self.src.len() {
            match self.src[self.i] {
                b' ' | b'\t' | b'\n' | b'\r' => self.i += 1,
                _ => break,
            }
        }
    }

    fn expect_word(&mut self, word: &str) -> JsonResult<()> {
        if self.text[self.i..].starts_with(word) {
            self.i += word.len();
            Ok(())
        } else {
            self.err(&format!("Expected '{}'", word))
        }
    }

    fn read_value(&mut self) -> JsonResult<Value> {
        self.skip_ws();
        if self.at_end() {
            return self.err("Unexpected end of JSON input");
        }
        match self.src[self.i] {
            b'{' => self.read_object(),
            b'[' => self.read_array(),
            b'"' => Ok(Value::Str(self.read_string()?)),
            b't' => { self.expect_word("true")?;  Ok(Value::Bool(true)) }
            b'f' => { self.expect_word("false")?; Ok(Value::Bool(false)) }
            b'n' => { self.expect_word("null")?;  Ok(Value::Null) }
            _    => self.read_number(),
        }
    }

    fn read_object(&mut self) -> JsonResult<Value> {
        let mut members: Vec<(String, Value)> = Vec::new();
        self.i += 1; // consume '{'
        self.skip_ws();
        if !self.at_end() && self.src[self.i] == b'}' {
            self.i += 1;
            return Ok(Value::Object(members));
        }
        loop {
            self.skip_ws();
            if self.at_end() || self.src[self.i] != b'"' {
                return self.err("Expected a string key");
            }
            let key = self.read_string()?;
            self.skip_ws();
            if self.at_end() || self.src[self.i] != b':' {
                return self.err(&format!("Expected ':' after key '{}'", key));
            }
            self.i += 1;
            let value = self.read_value()?;
            members.push((key, value));
            self.skip_ws();
            if self.at_end() {
                return self.err("Unterminated object");
            }
            match self.src[self.i] {
                b',' => { self.i += 1; }
                b'}' => { self.i += 1; return Ok(Value::Object(members)); }
                _ => return self.err("Expected ',' or '}'"),
            }
        }
    }

    fn read_array(&mut self) -> JsonResult<Value> {
        let mut items: Vec<Value> = Vec::new();
        self.i += 1; // consume '['
        self.skip_ws();
        if !self.at_end() && self.src[self.i] == b']' {
            self.i += 1;
            return Ok(Value::Array(items));
        }
        loop {
            items.push(self.read_value()?);
            self.skip_ws();
            if self.at_end() {
                return self.err("Unterminated array");
            }
            match self.src[self.i] {
                b',' => { self.i += 1; }
                b']' => { self.i += 1; return Ok(Value::Array(items)); }
                _ => return self.err("Expected ',' or ']'"),
            }
        }
    }

    fn read_hex4(&mut self) -> JsonResult<u32> {
        if self.i + 4 > self.src.len() {
            return self.err("Truncated \\u escape");
        }
        let mut cp: u32 = 0;
        for k in 0..4 {
            let c = self.src[self.i + k];
            let d = match c {
                b'0'..=b'9' => (c - b'0') as u32,
                b'a'..=b'f' => (c - b'a' + 10) as u32,
                b'A'..=b'F' => (c - b'A' + 10) as u32,
                _ => return self.err("Invalid \\u escape"),
            };
            cp = (cp << 4) | d;
        }
        self.i += 4;
        Ok(cp)
    }

    fn read_string(&mut self) -> JsonResult<String> {
        self.i += 1; // consume opening quote
        let mut out = String::new();
        loop {
            if self.at_end() {
                return self.err("Unterminated string");
            }
            let c = self.src[self.i];
            if c == b'"' {
                self.i += 1;
                return Ok(out);
            }
            if c != b'\\' {
                // Copy one whole UTF-8 character.
                let rest = &self.text[self.i..];
                let ch = match rest.chars().next() {
                    Some(ch) => ch,
                    None => return self.err("Unterminated string"),
                };
                out.push(ch);
                self.i += ch.len_utf8();
                continue;
            }
            self.i += 1; // consume backslash
            if self.at_end() {
                return self.err("Unterminated escape");
            }
            let e = self.src[self.i];
            self.i += 1;
            match e {
                b'"'  => out.push('"'),
                b'\\' => out.push('\\'),
                b'/'  => out.push('/'),
                b'b'  => out.push('\u{08}'),
                b'f'  => out.push('\u{0c}'),
                b'n'  => out.push('\n'),
                b'r'  => out.push('\r'),
                b't'  => out.push('\t'),
                b'u'  => {
                    let mut cp = self.read_hex4()?;
                    // Combine a surrogate pair when both halves are present.
                    if (0xD800..=0xDBFF).contains(&cp)
                        && self.i + 1 < self.src.len()
                        && self.src[self.i] == b'\\'
                        && self.src[self.i + 1] == b'u'
                    {
                        let save = self.i;
                        self.i += 2;
                        let lo = self.read_hex4()?;
                        if (0xDC00..=0xDFFF).contains(&lo) {
                            cp = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00);
                        } else {
                            self.i = save;
                        }
                    }
                    match char::from_u32(cp) {
                        Some(ch) => out.push(ch),
                        None => return self.err("Invalid code point in \\u escape"),
                    }
                }
                _ => return self.err("Invalid escape"),
            }
        }
    }

    fn read_number(&mut self) -> JsonResult<Value> {
        let start = self.i;
        if !self.at_end() && self.src[self.i] == b'-' {
            self.i += 1;
        }
        while !self.at_end() {
            match self.src[self.i] {
                b'0'..=b'9' | b'.' | b'e' | b'E' | b'+' | b'-' => self.i += 1,
                _ => break,
            }
        }
        if start == self.i {
            return self.err("Expected a value");
        }
        let text = &self.text[start..self.i];
        if text.parse::<f64>().is_err() {
            return Err(JsonError::new(format!(
                "Invalid number '{}' at offset {}", text, start)));
        }
        Ok(Value::Number(text.to_string()))
    }
}

pub fn parse(text: &str) -> JsonResult<Value> {
    let mut p = Parser::new(text);
    let v = p.read_value()?;
    p.skip_ws();
    if !p.at_end() {
        return Err(JsonError::new(format!("Trailing content at offset {}", p.i)));
    }
    Ok(v)
}

// ---------------------------------------------------------------------
// Field accessors
// ---------------------------------------------------------------------

fn type_error<T>(ctx: &str, expected: &str, actual: &Value) -> JsonResult<T> {
    Err(JsonError::new(format!(
        "JSON field '{}' is not {} (got {})", ctx, expected, actual.describe())))
}

pub fn require<'a>(obj: &'a Value, key: &str) -> JsonResult<&'a Value> {
    match obj {
        Value::Object(_) => obj
            .get(key)
            .ok_or_else(|| JsonError::new(format!("Missing JSON field '{}'", key))),
        _ => Err(JsonError::new(format!(
            "Expected an object holding field '{}'", key))),
    }
}

pub fn as_string(v: &Value, ctx: &str) -> JsonResult<String> {
    match v {
        Value::Str(s)    => Ok(s.clone()),
        Value::Number(n) => Ok(n.clone()),
        Value::Bool(b)   => Ok(if *b { "true".to_string() } else { "false".to_string() }),
        Value::Null      => Ok(String::new()),
        _ => type_error(ctx, "a string", v),
    }
}

pub fn as_f64(v: &Value, ctx: &str) -> JsonResult<f64> {
    match v {
        Value::Number(n) => n.parse::<f64>().map_err(|_| {
            JsonError::new(format!("JSON field '{}' is not a number", ctx))
        }),
        Value::Str(s) => s.trim().parse::<f64>().map_err(|_| {
            JsonError::new(format!("JSON field '{}' is not a number", ctx))
        }),
        _ => type_error(ctx, "a number", v),
    }
}

pub fn as_i32(v: &Value, ctx: &str) -> JsonResult<i32> {
    let d = as_f64(v, ctx)?;
    if d.fract() != 0.0 || d < i32::MIN as f64 || d > i32::MAX as f64 {
        return type_error(ctx, "an integer", v);
    }
    Ok(d as i32)
}

pub fn as_bool(v: &Value, ctx: &str) -> JsonResult<bool> {
    match v {
        Value::Bool(b) => Ok(*b),
        Value::Str(s) => {
            let low = s.trim().to_lowercase();
            match low.as_str() {
                "true"  | "t" | "yes" | "y" | "1" => Ok(true),
                "false" | "f" | "no"  | "n" | "0" => Ok(false),
                _ => type_error(ctx, "a boolean", v),
            }
        }
        _ => type_error(ctx, "a boolean", v),
    }
}

pub fn as_array<'a>(v: &'a Value, ctx: &str) -> JsonResult<&'a Vec<Value>> {
    match v {
        Value::Array(items) => Ok(items),
        _ => type_error(ctx, "an array", v),
    }
}
